builtin/credential/aws/path_role.go

```diff
@@ -319,6 +319,112 @@ func (b *backend) setRole(ctx context.Context, s logical.Storage, roleName strin
 	return nil
 }
 
+// initialize is used to initialize the AWS roles
+func (b *backend) initialize(ctx context.Context, req *logical.InitializationRequest) error {
+
+	// on standbys and DR secondaries we do not want to run any kind of upgrade logic
+	if b.System().ReplicationState().HasState(consts.ReplicationPerformanceStandby | consts.ReplicationDRSecondary) {
+		return nil
+	}
+
+	// Initialize only if we are either:
+	//   (1) A local mount.
+	//   (2) Are _NOT_ a replicated performance secondary
+	if b.System().LocalMount() || !b.System().ReplicationState().HasState(consts.ReplicationPerformanceSecondary) {
+
+		s := req.Storage
+
+		logger := b.Logger().Named("initialize")
+		logger.Info("starting initialization")
+
+		go func() {
+			// The vault will become unsealed while this goroutine is running,
+			// so we could see some role requests block until the lock is
+			// released.  However we'd rather see those requests block (and
+			// potentially start timing out) than allow a non-upgraded role to
+			// be fetched.
+			b.roleMutex.Lock()
+			defer b.roleMutex.Unlock()
+
+			upgraded, err := b.upgrade(ctx, s)
+			if err != nil {
+				logger.Error("error running initialization", "error", err)
+				return
+			}
+			if upgraded {
+				logger.Info("an upgrade was performed during initialization")
+			}
+
+			logger.Info("initialization succeeded")
+		}()
+
+	}
+
+	return nil
+}
+
+// roleStorageVersion stores the latest role storage version that we have
+// upgraded to.
+type roleStorageVersion struct {
+	Version int `json:"version"`
+}
+
+// upgrade and persist all of the role entries that are in need of being
+// upgraded.
+func (b *backend) upgrade(ctx context.Context, s logical.Storage) (bool, error) {
+
+	// find the current role-storage-version
+	version := -1
+	entry, err := s.Get(ctx, "config/role-storage-version")
+	if err != nil {
+		return false, err
+	}
+	if entry != nil {
+		var rsv roleStorageVersion
+		err = entry.DecodeJSON(&rsv)
+		if err != nil {
+			return false, err
+		}
+		version = rsv.Version
+	}
+
+	// check if we need to upgrade
+	if version == currentRoleStorageVersion {
```

Comment:
Please follow the model used in the role upgrade function, where it uses a switch with fallthroughs to determine whether to upgrade.

Reply:
I made it clearer what was actually happening there, that for now we upgrade only for role version, but in the future we may add more criteria.